mrp_merge_productions/mrp_production.py

```diff
@@ -0,0 +1,320 @@
+#    (<http://www.savoirfairelinux.com>).
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp.osv import orm
+from tools.translate import _
+
+import logging
+_logger = logging.getLogger(__name__)
+
+
+def id_of(obj):
+    """ For o2m relations, returns the foreign id if it is set or None """
+    if obj:
+        return obj.id
+    else:
+        return None
+
+
+def get_target(productions):
+    def sort_key(prod):
+        return (
+            # Prioritize ready productions
+            {"ready": 1}.get(prod.state, 9),
+            # Then the oldest id
+            prod.id,
+        )
+
+    productions = sorted(productions, key=sort_key)
+    return productions[0], productions[1:]
+
+
+def name_of(obj):
+    return getattr(obj, "name", repr(obj))
+
+
+def moves_mergeable(source, target):
+    return all((
+        target.state not in ('cancel', 'done', 'in_production'),
+        source.product_id.id == target.product_id.id,
+        id_of(source.prodlot_id) == id_of(target.prodlot_id),
+        id_of(source.product_uom) == id_of(target.product_uom),
+        source.location_id.id == target.location_id.id,
+        source.location_dest_id.id == target.location_dest_id.id,
+    ))
+
+
+def lines_mergeable(source, target):
+    return source.product_id.id == target.product_id.id
+
+
+def merge_prodlines(obj, cr, uid, source_lines, target_lines, updates,
+                    to_delete):
+    for to_merge in source_lines:
+        for mergeable in target_lines:
+            if lines_mergeable(to_merge, mergeable):
+                to_write = updates.setdefault(
+                    mergeable.id,
+                    {"product_qty": mergeable.product_qty,
+                     "product_uos_qty": mergeable.product_uos_qty}
+                )
+                _logger.debug("Merging prodline %d into %d (%d %s)",
+                              to_merge.id, mergeable.id,
+                              to_merge.product_qty, to_merge.product_id.name)
+                to_write["product_qty"] += to_merge.product_qty
+                if to_merge.product_uos_qty:
+                    to_write["product_uos_qty"] += to_merge.product_uos_qty
+                to_delete.append(to_merge.id)
+                break
+        else:
```

Comment:
Is the indentation right here ? Looks like a weird "for ...: / else: " construct.